### PR TITLE
feat: adds script for git migration.

### PR DESCRIPTION
**Pull request by atulep**
Script that migrates Git history of a split repo to google-cloud-python.

It copies Git history from source repo (i.e. "split repo") to target repo (i.e. `google-cloud-python`). Then, creates a pull request in target repo.

### CODE REVIEW

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/run.sh:48]: You should set `$UPDATE_SCRIPT` for use by the `git_migrate_history` to point to my script in #10880 . Or pass it as a parameter.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:131]: LMK whether you want to commit before calling `$UPDATE_SCRIPT`, which would result in 2 commits per migration, or have everything in one commit.

If the latter, let me know whether you want to commit here; I'm currently committing at the end of `$UPDATE_SCRIPT`

File: scripts/split_repo_migration/run.sh
Comment (atulep) [reply scripts/split_repo_migration/run.sh:48]: Done. Pass as a parameter.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:131]: I think having two commits help with debugging and potential rollbacks.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:45]: IIUC, this could be an arbitrary path in the file system, so it could have spaces we should enclose it in quotes: `"$8"`. In most cases, it's advisable to quote shell variable references for safety.

Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:45]: Is it true that we normally provide the `UPDATE_SCRIPT` parameter but may not always provide `BRANCH`? In that case, it may be advisable to make `UPDATE_SCRIPT` be `$7` and `BRANCH` be `$8`, so it's easier to omit. (otherwise, we have to pass `''` for `$7 ` in order to be able to pass something to `$8`.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:20]: Update usage comment to include `UPDATE_SCRIPT`

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:49]: quote RHS

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:90]: observation: I'm pleasantly surprised this works without quotes. In the shell `FOO=alpha beta` differs from `FOO="alpha beta"`, but it seems like `FOO=$(echo "alpha beta")` works like the latter.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:101]:
```suggestion
    for file in "${KEEP_FILES_SPACES[@]}"
    do
      KEEP_FILE_COMMANDS+="; git checkout -- ${file} 2>/dev/null || true"
    done
```

Isn't this simpler, putting the semicolon separator at the start of the line? Less code duplication.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:61]: We're fine not quoting for most likely cases, but `mktemp` uses `$TMPDIR` if set, and if that has spaces, then we would need to quote the RHS here as well as all references to `${WORKDIR}`

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:142]:
```suggestion
# Do a diff between source code split repo and migrated code.
```

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:134]: I'm realizing I wasn't thinking about things correctly and used the wrong language here, so let me try to be accurate now.

This script pushes all the split-repo's commits (git commits) to the monorepo branch `$BRANCH`. So we have hundreds of commits, not just one or two.

IIUC (correct me if I'm wrong!), we have two options:

(A) We could issue a PR at this point. IIUC, the PR would simply merge the `$BRANCH` with this whole imported history into the `main` branch. Then after that's merged, we run my postprocessing script on the `main` branch directly and issue a PR against the `main` branch. This means that between the two PRs, there's a period where the imported split-repo shows up as a package dir in the monorepo `main`, but hasn't been fully migrated because the post-processing script hasn't run.

or

(B) On this branch `$BRANCH`, after all these imported commits, we run my post-process script and add one final commit with those changes. Then we issue a single PR to merge the whole branch into `main`. When it is merged, the new package directory will have all the migration changes it needs; we won't ever be in a state where the repo dir appears migrated but isn't really. If we want to review just the post-processing changes, we merely look at the last commit in the PR.

Writing this out, I think option B would be cleaner. WDYT? Please let me know whether there are aspects I'm failing to consider.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:21]: If the user sources this script rather than executing it, this will kill their shell. I suggest at the top of the file:
- either enforcing that the script is being executed, or
- defining `$EXIT` and using it in place of `exit`, like so:
  ```
  [[ "${BASH_SOURCE[0]}" != "${0}" ]] && EXIT=return || EXIT=exit
  ```

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:134]: Talked about this offline - the path forward here is to apply an update script in batch on all split repos after they have been migrated. This results in two PRs.

Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:134]: Removed update script.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:45]: Removed update script.

Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:45]: Removed update script.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:20]: N/A

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:49]: Removed update script.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:21]: Nice catch! thanks.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:95]: Oh, interesting. It doesn't error on `zsh`, which I think makes sense. Code duplication makes me nervous. I guess maybe it's OK because we only use this script during the migration, and the two usages are easy to compare.

The really right solution, I suppose, is to write all the commands, including these, that currently go into `${FILTER}` into an executable file like  `/tmp/cleantree.sh` instead (one per line, so no need to worry about semicolons nor, hence, repetition). Then below we could use  `git filter-branch --tree-filter /tmp/cleantree.sh`

I'll leave it up to you. If we wind up not needing to modify this logic, I think leaving the code as-is is OK for speed. But if we wind up tweaking it, using a file might be better.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:139]: Might be nice to redirect the `diff` output to a file and if the `diff` returns an error code (diff not empty) print the name of the file here so people can easily inspect.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:145]: Do we want to go ahead and uncomment this here, or do we want the caller to do the git push and PR creation? Either is fine. If we have the caller do it, I'll include it in the batch-processing script.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:20]: We're accepting up to 7 arguments, correct?

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/run.sh:21]: Could you insert the source vs exec guard here too so we don't accidentally kill the user's shell?

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/run.sh:32]: remove cruft?

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/run.sh:27]: We were talking about deriving the `$ARTIFACT_NAME` from the split repo, possibly using `jq`. I think that's a good idea (single source of truth, less chance for inconsistencies), though I don't know whether you want to do it here or in `git-migrate-history`.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:145]: You'll need access to tempdir where the generated code got placed in. I think it's easier to push from my script since I have that info here.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:20]: Correct. The last 3 are optional though.

File: scripts/split_repo_migration/run.sh
Comment (atulep) [reply scripts/split_repo_migration/run.sh]: Done

File: scripts/split_repo_migration/run.sh
Comment (atulep) [reply scripts/split_repo_migration/run.sh]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:20]: Oh, I see.  My bad.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:142]:
```suggestion
DIFF_FILE="${WORKDIR}/diff.txt"
if diff -r "target-repo/${TARGET_PATH}" source-repo-validation > "${DIFF_FILE}" ; then
  echo "No diff"
else
  echo "Diff non-empty. See ${DIFF_FILE}"
  $EXIT 1
fi
```

I think the above  should work; the redirection is separate from the exit code.

You could even shorten this as follows if you want:

```suggestion
DIFF_FILE="${WORKDIR}/diff.txt"
[[ diff -r "target-repo/${TARGET_PATH}" source-repo-validation > "${DIFF_FILE}" ]]  \
  &&   echo "No diff" \
  || { echo "Diff non-empty. See ${DIFF_FILE}" ;  $EXIT 1 ; }

```

**Review by vchudnov-g — COMMENTED** [review]
One relatively minor comment.

But I do think we want the other open comment, deriving `${ARTIFACT_NAME}` from the split repo.

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/run.sh:51]: So this assumes we're running from a directory that contains subdirectories
`googleapis/${SPLIT_REPO}/` and `googleapis/google-cloud-python` Would be good to note this in usage comments at the top, but as long as I know I can make the batch migration script conform to this.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:121]: We should talk. I propose you don't clone $TARGET_REPO, but allow it to be provided as a parameter to run.sh, and here you just assume it exists. If you do this, then in my batch update script I can clone $TARGET_REPO once, and use `git worktree` to create a new workdir directory referencing that cloned repo with an API-specific branch created from `main`. Then the rest of this script can import the changes in that directory. 

This way we can easily both generate multiple per-API merge PRs simumltaneously AND not commit google-cloud-python repeatedly. WDYT? Happy to talk more.

File: scripts/split_repo_migration/run.sh
Comment (vchudnov-g) [reply scripts/split_repo_migration/run.sh:29]: Related: assuming you can derive `$ARTIFACT_NAME`, than that won't need to be `$2`. Instead, make `$2` an optional branch name that you pass to `git-migrate-history` as `$7`

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:125]: The branch may already exist; this is what I'm envisioning for the batch migration script. So we shoul dmake sure this doesn't error. I think using `-B` instead of `-b` should make this case work (uses an existing branch if it exists, or creates a new one)

Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Done.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Sounds good. I no longer clone TARGET_REPO and instead accept path to target repo as an argument.

File: scripts/split_repo_migration/run.sh
Comment (atulep) [reply scripts/split_repo_migration/run.sh]: N/A. Removed run.sh

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Thanks for suggestion. Added.

File: scripts/split_repo_migration/run.sh
Comment (atulep) [reply scripts/split_repo_migration/run.sh]: Removed artifcat name.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:42]: nit: Update the "format" comment on the previous line

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:115]: We need to prepend `packages/` here, don't we? It's not included [in the json file](https://github.com/googleapis/python-speech/blob/4c2a6fd82d66fbbe2ddc24e222a2415ecd8def20/.repo-metadata.json#L11) and we need to conform to [this directory structure](https://github.com/googleapis/google-cloud-python/tree/main/packages)

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:125]: Curious: Aren't all these commands (after the `shopt`) equivalent to just this?:
```
mkdir -p ${TARGET_PATH}; mv * ${TARGET_PATH}
```

Why do move stuff into `migrated-source` just to immediately move it out? It's innocuous to leave it in, I guess, but does this have any side effects I'm not seeing?

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:134]: Since we'll be working with worktrees, the remotes are shared. We should uniquify them via
```
REMOTE="remote.${SOURCE_REPO}"
...
git merge --allow-unrelated-histories ${REMOTE}/main --no-edit
```

Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:134]: If `${SOURCE_REPO}` doesn't work we can use `${BRANCH}` instead.

Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Used SOURCE_REPO

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:125]: Not sure, I copied this from Node.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Nice catch. Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:118]: This will always evaluate to true, so we should just take out the conditional. Or we can have the conditional on just the output of `jq`.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:157]: I assume you didn't mean to leave these lines commented out, and the plan is still for this script to create the PR?

**Review by vchudnov-g — APPROVED** [review]
Looks good. Only issue is about the commented-out `gh` command. Could you address that before merging?

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:162]: This command was prompting me for PR info, which we don't want if we run this in a batch script. Let's figure out how to pass everything on the command line.

Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:162]: We could do this as a follow-up PR, IMO.

Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:162]: I think just adding `--body "Migrate ${SRC_API_NAME} into ${TARGET_PATH}"` should do it.

Comment (vchudnov-g) [reply scripts/split_repo_migration/git-migrate-history.sh:162]: Also, we need a conventional commit title format....

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:155]: As per offline chat, sounds like we also need to `mkdir -p owl-bot-staging/${DISTRIBUTION_NAME}` after validation and before creating the PR.

Which means maybe we do the `jq` extract into `${DISTRIBUTION_NAME}` and then set `TARGET_PATH="packages/${DISTRIBUTION_NAME}"` above?

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:115]: We might need the result of the query separate from the path. See comment below about `owl-bot-staging/`

**Review by vchudnov-g — CHANGES_REQUESTED** [review]

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Uncommented.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:56]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Done

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh:155]: We should add any necessary fixes, if needed, in follow-up PRs.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (vchudnov-g) [inline scripts/split_repo_migration/git-migrate-history.sh:168]: Doesn't the conventional commit need to appear in the title? For my sample PR this morning, I wound up manually changing the title to `chore(migration): Import googleapis/python-ids into packages/google-cloud-ids` and that passed the presubmit checks. Should we try something like that?

**Review by vchudnov-g — APPROVED** [review]
One non-trivial comment about making the *title* have a conventional  commit format.

File: scripts/split_repo_migration/git-migrate-history.sh
Comment (atulep) [reply scripts/split_repo_migration/git-migrate-history.sh]: Yeah, I updated title. It will be something like:

```chore: Import googleapis/python-ids into packages/google-cloud-ids (#11104)

See #11104.
```

**Review by vchudnov-g — APPROVED** [review]

**Pull request merged**
